Editorial review: Document scroll-triggered animations (#43158)
mdn/content: files/en-us/web/css/reference/properties/animation-trigger/index.md

animation-trigger: none;

/* Dashed ident plus a single <animation-action> */
animation-trigger: --my-trigger play;
animation-trigger: --my-other-trigger play-once;

/* Dashed ident plus two <animation-action>s */
animation-trigger: --my-trigger play-forwards play-backwards;
animation-trigger: --my-other-trigger play reset;

/* Multiple animation-trigger values */
animation-trigger:
  none,
  --my-trigger play-forwards play-backwards,
  --my-other-trigger play reset;

/* Global values */
animation-trigger: inherit;
animation-trigger: initial;
animation-trigger: revert;
animation-trigger: revert-layer;
animation-trigger: unset;
```

The `animation-trigger` property may be specified using one or more `animation-trigger` values separated by commas. Each `animation-trigger` value consists of the keyword `none`, or a {{cssxref("dashed-ident")}} followed by one or two {{cssxref("animation-action")}} values.

### Values

- `none`
  - : Specifies that the animation is not a scroll-triggered animation.
- {{cssxref("dashed-ident")}}
  - : An identifier equal to the {{cssxref("timeline-trigger-name")}} of the element that will trigger the animation.
- {{cssxref("animation-action")}}

>   - : Specifies the behavior of the animation once activated (in the case of the first value) and deactivated (in the case of the second, optional value).

Reviewer: ```suggestion
  - : Specifies the behavior of the animation once the trigger is activated (in the case of the first value) and deactivated (in the case of the second, optional value).
```

  Etiam a tincidunt purus. Maecenas semper sed enim at blandit. Aenean ut
  sagittis lorem, eget gravida purus. Phasellus eleifend, lectus nec pulvinar
  facilisis, dui dolor feugiat odio, iaculis tempor felis est non tortor. In
  suscipit lorem efficitur molestie tempus. Integer sit amet neque et risus
  iaculis sodales sed eget diam. Quisque sodales nunc sapien, vitae lacinia ex
  luctus quis. Maecenas scelerisque scelerisque elit eu consequat. Etiam ac
  tristique tellus, sed tincidunt velit.
</p>
```

#### CSS

The {{htmlelement("div")}} element has an `animation` applied that smoothly inverts its colors. We set an `animation-trigger` value on it that references a `timeline-trigger-name` of `--t`; we also specify two `<animation-action>` values — `play-forwards` and `play-backwards` — which specify that the animation will play forwards on activation, and play in reverse on deactivation.

We then specify a `timeline-trigger` value of `--t view() contain` on the same `<div>`, meaning that it is the trigger for its own animation. This value includes the identifier referenced in the `animation-trigger` property, plus:

> - A `timeline-trigger-source` value of [`view()`](/en-US/docs/Web/CSS/Reference/Properties/animation-timeline/view), which sets the element providing the view progress timeline as the nearest scrolling ancestor element.

Reviewer: ```suggestion
- A `timeline-trigger-source` value of [`view()`](/en-US/docs/Web/CSS/Reference/Properties/animation-timeline/view), which sets the source of the trigger to the view progress timeline tracking the element inside its nearest scrolling ancestor element.
```

/* Multiple values */
animation-trigger:
  none,
  -forwards play-backwards,
  --my-other-trigger play reset;

/* Global values */
animation-trigger: inherit;
animation-trigger: initial;
animation-trigger: revert;
animation-trigger: revert-layer;
animation-trigger: unset;
```

### Values

The property accepts a comma-separated list of values. Each value is either the keyword `none` or a {{cssxref("dashed-ident")}} followed by one or two {{cssxref("animation-action")}} values.

- `none`
  - : The associated animation is not a triggered animation.
- {{cssxref("dashed-ident")}}
  - : A custom identifier defining the name of the trigger that will trigger the animation.
- {{cssxref("animation-action")}}
  - : An `<animation-action>`: The keywords `none`, `play`, `play-forwards`, `play-backwards`, `play-once`,`pause`, `replay`, or `reset`.

## Description

The `animation-trigger` property specifies which trigger will control an animated element's animations. A value other than `none` turns the animation into a [scroll-triggered animation](/en-US/docs/Web/CSS/Guides/Animation_triggers/Using_scroll-triggered_animations).

### Defining a trigger

> The trigger is identified via a `<dashed-ident>` value, which is defined in the tracked element's {{cssxref("timeline-trigger-name")}} property. If there is no scrolling element with the same `<dashed-ident>` set as the `timeline-trigger-name` value, the animation will not have a trigger and will play as soon as the page loads.

Reviewer 2: if there is no animation-trigger set, the animation plays on load...
it's if there is an animation-trigger set, it will only play if triggered, so if there is no actual trigger, the animation can't happen.

This is saying the animation will happen.

Author: Ah yes, I got confused here!

I've fixed this in my latest commit.

```css
.trigger {
  timeline-trigger: --my-trigger view();
}
```

If an element has an animation _and_ an `animation-trigger` set on it, but no scrolling element exists with the same `<dashed-ident>` set as its `timeline-trigger-name` value, the animation will not have a trigger and therefore will never play.

### Defining the triggered animation actions

The `animation-trigger` value must include one or two {{cssxref("animation-action")}} keywords after the `<dashed-ident>` to specify the behavior of the animation when the trigger is activated and deactivated. If two `<animation-action>`s are specified, the first is the activation action and the second is the deactivation action. If only one `<animation-action>` is set, this is the activation action, and there is no deactivation.

For example:

```css
.animated {
  animation: rotate 3s infinite linear both;
  animation-trigger: --my-trigger play-forwards play-backwards;
}
```

When the trigger is activated, the animation will `play-forwards`. When the trigger is deactivated, the animation will `play-backwards`.

There are eight `<animation-action>` values, each providing different animation behaviors.

> Setting `play-forwards play-backwards` is a common pattern, often used to "animate in" an element when its trigger scrolls into view, and then "animate out" the element again when the trigger scrolls out of view.

Reviewer 2: I think our readers will take a long time to grok "active" and "inactive" ( as well as the fact that the action triggers a trigger, rather than directly triggering an animation). Thanks for the suggestion. we do need to be pedantically correct. How about a combination? It's wordier, but possibly easier to understand, and may even help readers understand active and inactive faster.

```suggestion
Setting `play-forwards play-backwards` is a common pattern, often used to "animate in" an element when its trigger becomes active, such as by scrolling into view, and then "animate out" the element again when the trigger becomes inactive, for example, by scrolling out of view.
```

Author: I think this makes sense.

> [!NOTE]
> It is possible for the animated element and the element that creates the timeline to be the same element.

### Triggering the same animation via multiple different triggers

If you have an animated element, and you want to define triggers on multiple different elements that all trigger the same animation, you need to specify the animation multiple times on the animated element, and then give each animation instance a different trigger.

For example:

```css
.animated {
  animation:
    moveright 2s 1 ease-out both,
    moveright 2s 1 ease-out forwards;
  animation-trigger:

>     --t play-forwards play-backwards,

Reviewer 2: ```suggestion
    --t1 play-forwards play-backwards,
```

When the trigger is activated, the animation will `play-forwards`. When the trigger is deactivated, the animation will `play-backwards`.

There are eight `<animation-action>` values, each providing different animation behaviors.

> Setting `play-forwards play-backwards` is a common pattern, often used to "animate in" an element when its trigger scrolls into view, and then "animate out" the element again when the trigger scrolls out of view.

Reviewer: It’s technically not correct to say that a _“trigger scrolls into view”_. The trigger can be controlled by a `{Scroll|View}Timeline` (currently the only type of trigger) but, in the the future, can also be something like an interaction.

```suggestion
Setting `play-forwards play-backwards` is a common pattern, often used to "animate in" an element when its trigger becomes active, and then "animate out" the element again when the trigger becomes inactive.
```

> The `play-once` action is generally used on its own or as part of `play-once pause`; setting `play-once` as the activation action causes the animation to play only once when it scrolls into view. The addition of `pause` on deactivation pauses the animation when the trigger scrolls out of its activation range, restarting from where it was paused if re-activated.

Reviewer: Same remark about the trigger “scrolling into view”

}
```

#### Result

The rendered result looks like this:

{{EmbedLiveSample("same-element", "100%", "240")}}

Try scrolling the content up. When the `<div>` fully appears in the scrollport, its animation will play; when any part of the `<div>` leaves the scrollport at either edge, the animation will play backwards.

### Multiple triggers for the same animation

In this example, we show how to assign multiple triggers to control the same animation. This example is very similar to our initial [basic usage example](#basic_usage), except that it has multiple trigger elements.

> 

Reviewer 2: ```suggestion

```html
...
<div class="trigger1">I create a trigger</div>
...
<div class="trigger2">I create another trigger</div>
...
<div class="trigger3">I create yet another trigger</div>
...
\```
```

Author: I've added this via my latest commit.

<p>
  Fusce dictum ex quis ipsum consectetur placerat. Cras sed lectus ex. Quisque
  purus dolor, vulputate ac mi eget, commodo varius odio. Suspendisse faucibus
  ipsum vel libero finibus, in placerat nibh congue. Sed iaculis, metus et
  euismod posuere, mi diam vestibulum felis, ac vulputate eros ipsum id justo.
  Etiam a tincidunt purus. Maecenas semper sed enim at blandit. Aenean ut
  sagittis lorem, eget gravida purus. Phasellus eleifend, lectus nec pulvinar
  facilisis, dui dolor feugiat odio, iaculis tempor felis est non tortor. In
  suscipit lorem efficitur molestie tempus. Integer sit amet neque et risus
  iaculis sodales sed eget diam. Quisque sodales nunc sapien, vitae lacinia ex
  luctus quis. Maecenas scelerisque scelerisque elit eu consequat. Etiam ac
  tristique tellus, sed tincidunt velit.
</p>
```

```css hidden live-sample___multiple-triggers
body {
  width: 80%;
  margin: 0 auto;
  font-family: Arial, Helvetica, sans-serif;
  font-size: 1.3rem;
}

div {
  height: 100px;
  border: 5px solid black;
}

.animated {
  width: 100px;
  background: orange;
}

.trigger1,
.trigger2,
.trigger3 {
  background: wheat;
}
```

> This time, our `.animated` `<div>` has the `rotate` `animation` applied to it three times. It then has three differently-named `animation-trigger` values set on it, one for each instance of the `animation`.

Reviewer 2: ```suggestion
This time, our `.animated` `<div>` has the same `rotate` animation applied to it three times using the `animation` shorthand property. It then has three differently-named `animation-trigger` values set on it, one for each animation instance.
```

```css live-sample___multiple-triggers
div.animated {
  animation:
    rotate 3s infinite linear both,
    rotate 3s infinite linear forwards,
    rotate 3s infinite linear forwards;
  animation-trigger:
    --t1 play-forwards play-backwards,
    --t2 play-forwards play-backwards,
    --t3 play-forwards play-backwards;
}
```

> This example contains three trigger `<div>` elements. On each one, we define a timeline trigger with a different name. These names correspond to the names referenced in the `.animated` `<div>` element's `animation-trigger` property.

Reviewer 2: ```suggestion
On each trigger `<div>` element we define a timeline trigger with a different name. These names correspond to the names referenced in the `.animated` `<div>` element's `animation-trigger` property.
```
